### ISSUE TITLE
benchmarking/profiling

### ISSUE THREAD

**Issue by stevengj** [JuliaStrings/utf8proc#12]
It would be good to perform some benchmarking of utf8proc against ICU, and in general to perform some profiling to see if there are any easy targets for optimization.

**Comment by ScottPJones** [comment]
>This is the case in Julia, because UTF-8 strings are validated when they are created.

@stevengj I agree that having valid UTF-8 input could be used to make `utf8proc_iterate` faster, unfortunately, that's not correct currently in Julia (although @JeffBezanson [I think] and I would like to make that true). About `libunistring`, pretty please avoid anything with GPL (except in packages), it makes Julia useless (or even dangerous) for people like me, trying to use Julia in a commericial project.  For really frequent operations, I'll see if I can get this up to ICU speeds...

**Comment by tkelman** [comment]
The author of utf8rewind pings me once in a while on reddit to let me know he's been adding new features, so this would be another thing to compare against eventually: https://bitbucket.org/knight666/utf8rewind/overview

It's MIT licensed and pretty small like utf8proc is, so we could borrow anything that turned out to be worth using.

**Comment by xhochy** [comment]
In the Apache Arrow project, we also evaluating using `utf8proc` and are running some benchmarks, see https://github.com/apache/arrow/pull/7449#issuecomment-645276441 Currently `unilib` seems to perform better for us.

**Comment by stevengj** [comment]
One could probably speed up upper/lowercase conversions, e.g. by adding a specialized table just for this, but it's not clear from the issue whether that functionality is actually performance critical or just a test case?

**Comment by stevengj** [comment]
This thesis claims to have implemented similar functionality with a considerable speedup:  https://bearworks.missouristate.edu/theses/2731/

However, the source code does not seem publicly available.  The thesis advisor is [tragically deceased](https://www.legacy.com/obituaries/name/eric-shade-obituary?pid=174514292), and I'm not sure about the contact information for the author.

**Comment by sgllama** [comment]
> This thesis claims to have implemented similar functionality with a considerable speedup: https://bearworks.missouristate.edu/theses/2731/
> 
> However, the source code does not seem publicly available. The thesis advisor is [tragically deceased](https://www.legacy.com/obituaries/name/eric-shade-obituary?pid=174514292), and I'm not sure about the contact information for the author.

This may be the thesis author: https://www.linkedin.com/in/jpdurham
